libflux/src/core/semantic/types.rs

```diff
@@ -663,6 +671,57 @@ impl Array {
     }
 }
 
+#[derive(Debug, Clone, PartialEq, Serialize)]
+pub struct Dictionary {
+    pub key: MonoType,
+    pub val: MonoType,
+}
+
+impl fmt::Display for Dictionary {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "[{}:{}]", self.key, self.val)
+    }
+}
+
+impl Substitutable for Dictionary {
+    fn apply(self, sub: &Substitution) -> Self {
+        Dictionary {
+            key: self.key.apply(sub),
+            val: self.val.apply(sub),
+        }
+    }
+    fn free_vars(&self) -> Vec<Tvar> {
+        union(self.key.free_vars(), self.val.free_vars())
+    }
+}
+
+impl MaxTvar for Dictionary {
+    fn max_tvar(&self) -> Tvar {
+        vec![self.key.max_tvar(), self.val.max_tvar()].max_tvar()
+    }
+}
+
+impl Dictionary {
+    fn unify(
+        self,
+        actual: Self,
+        cons: &mut TvarKinds,
+        f: &mut Fresher,
+    ) -> Result<Substitution, Error> {
+        let sub = self.key.unify(actual.key, cons, f)?;
+        apply_then_unify(self.val, actual.val, sub, cons, f)
```

Comment:
These 2 lines are the crux of the PR. The rest is either boilerplate or generated code.